### PR TITLE
Get tags into a separate structure that can be applied to other models

### PR DESCRIPTION
**Pull request by gabelula**
## What does this PR do?

PT Story: https://www.pivotaltracker.com/story/show/142993479

Gets tags out of the Comment Model and into its own. We still need to have something in the FE to show the tags only to intended roles, look on how the system of decisions on who can apply them will work, add tags to the User Model.

## How do I test this PR?

- Everything should be working as before. We only changed on how the tags are being implemented in the backend.

### CODE REVIEW

**Review by wyattjoh — CHANGES_REQUESTED** [review]
Tags should live on the models themselves, as it would act like metadata instead of splitting it into a separate collection. I would advise adjusting this to instead behave like our status_history behaves where we just apply atomic operations to the documents to add new tags and remove them.

I think this introduces too many new factors.

File: models/comment.js
Comment (wyattjoh) [inline models/comment.js]: The tags id should not be a part of this index, it is then coupling the id of the comment with the id of the tag, which doesn't make sense.

File: models/user.js
Comment (wyattjoh) [inline models/user.js]: Should remove this commented out code.

File: services/comments.js
Comment (wyattjoh) [inline services/comments.js]: There's a few issues here, I'll push some changes that resolve some issues with promises + query paths.

File: services/comments.js
Comment (wyattjoh) [inline services/comments.js]: If you are just performing an update, then just use the update instead of the `findOneAndUpdate`. Also, there is a `nMatched` that should be used to check to see if the query found something.

File: test/server/graph/loaders/metrics.js
Comment (wyattjoh) [inline test/server/graph/loaders/metrics.js]: An extra space here should be removed.

File: test/server/graph/mutations/removeCommentTag.js
Comment (wyattjoh) [inline test/server/graph/mutations/removeCommentTag.js]: Should use `await` here, it escapes context by not calling return/await and we won't capture the error.

File: test/server/graph/mutations/removeCommentTag.js
Comment (wyattjoh) [inline test/server/graph/mutations/removeCommentTag.js]: All database operations should be in a `beforeEach/before/it` block, otherwise it's not scoped properly for tests.

File: test/server/services/comments.js
Comment (wyattjoh) [inline test/server/services/comments.js]: Same as before, need `await/return` to resolve.

**Review by wyattjoh — CHANGES_REQUESTED** [review]

**Review by wyattjoh — DISMISSED** [review]
After performing a review of the backend, I've made some alterations to the schema to resolve some issues with the workflow. The frontend should be adjusted to match the new returns and graph structure in the `typeDefs.graphql` file.

**Comment by jde** [comment]
The permissions logic and tagLink data looks right. I can see how this separates the concept of tags cleanly for both tags that we want to store in a list as well as ad hoc tags!

The add/remove mutations, however, are still tightly bound to comments. Looking into the functions, the only thing that makes this comment specific is which service is used to add the tag.

I'd like to change addCommentTag, etc.. to addTag, etc... and pass in which model we want to add the tag to. This should realize our flexible tagging architecture for all product needs going forward.

File: graph/mutators/comment.js
Comment (jde) [inline graph/mutators/comment.js]: Change addCommentTag to addTag. Accept input concerning which model to add the tag to.

Comment (jde) [reply graph/mutators/comment.js]: My recommendation would be to create a thin getServiceFromString function that takes a string and returns a the service object. We can then check to make sure that service contains `addTag()` and use that interchangeably below.

File: graph/mutators/comment.js
Comment (jde) [inline graph/mutators/comment.js]: `COMMENTS` here would become dynamic based on input from the query variables

File: graph/mutators/comment.js
Comment (jde) [inline graph/mutators/comment.js]: `CommentsService` here and below would be the service returned by `getServiceFromString` at the outset of this function.

File: graph/mutators/comment.js
Comment (jde) [inline graph/mutators/comment.js]: See notes from addCommentTag.

File: graph/resolvers/root_mutation.js
Comment (jde) [inline graph/resolvers/root_mutation.js]: `addCommentTag` becomes `addTag`, passes through `model` as part of input

File: graph/typeDefs.graphql
Comment (jde) [inline graph/typeDefs.graphql:92]: We can create something like this:

```
enum TaggableModels {
  COMMENT,
  USER,
  ASSET
}
```

To control which models can receive comments.

File: graph/typeDefs.graphql
Comment (jde) [inline graph/typeDefs.graphql]: See above recommendations to allow this query to apply to any `TaggabeModel`.

File: services/comments.js
Comment (jde) [inline services/comments.js]: This function and `removeTag` live in `services/tags.js` and be imported and appended to all services for 'taggable" models.

**Review by jde — CHANGES_REQUESTED** [review]
This review contains a plan to introduce a concept "taggable" that encapsulates the tagging logic in a way that can be applied to any model.

File: graph/resolvers/util.js
Comment (jde) [inline graph/resolvers/util.js]: This should be ADMIN || MODERATOR atm. Are we going to a user.canDo('seeTags') type of scheme to centralize all role logic on the backend or does it make the most sense to explicitly hardcode it here?

**Review by jde — CHANGES_REQUESTED** [review]

File: graph/resolvers/util.js
Comment (wyattjoh) [reply graph/resolvers/util.js]: Makes sense to me, will implement those changes.

**Review by jde — APPROVED** [review]

**Comment by jde** [comment]
Back end implementation looks good to me.

File: errors.js
Comment (alinhle) [inline errors.js]: Looks like the copy and paste devil was here ;-)

**Comment by okbel** [comment]
@wyattjoh Frontend functionality ready! Please check, best and staff tag.

**Review by jde — APPROVED** [review]

**Comment by kgardnr** [comment]
Okay @wyattjoh @okbel @jde, we can merge this once the conflicts are resolved!

File: client/coral-framework/graphql/mutations.js
Comment (alinhle) [inline client/coral-framework/graphql/mutations.js]: Here should be `...AddTagResponse` to make it extendable.

File: client/coral-framework/graphql/mutations.js
Comment (alinhle) [inline client/coral-framework/graphql/mutations.js:232]: Here should be `...RemoveTagResponse` to make it extendable.

**Review by alinhle — CHANGES_REQUESTED** [review]
- `Response` interface has been changed. The plugins that defines new responses must follow this or the break (like reactions).
- graphql framework has some wrong assumptions that need to be removed.

**Review by alinhle — APPROVED** [review]
GTG after wyatt approves!

**Pull request merged**
